packages/actor-rdf-resolve-quad-pattern-federated/lib/BlankNodeSkolemizable.ts

```diff
@@ -0,0 +1,22 @@
+import * as RDF from "rdf-js";
+
+export class BlankNodeSkolemizable implements RDF.BlankNode {
+  public readonly termType: 'BlankNode' = 'BlankNode';
+  public readonly value: string;
+  /**
+   * This value can be obtained by consumers in query results,
+   * so that this can be passed into another query as an IRI,
+   * in order to obtain more results relating to this (blank) node.
+   */
+  public readonly skolemized: RDF.NamedNode;
+
+  constructor(value: string, skolemized: RDF.NamedNode) {
+    this.value = value;
+    this.skolemized = skolemized;
+  }
+
+  public equals(other: RDF.Term | null | undefined): boolean {
+    return other && other.termType === 'BlankNode' && other.value === this.value;
```

Comment:
I expect this to be the default implementation?

That said, we need to think whether this is the right implementation. On the one hand, we don't want to mess with symmetry and transitivity. On the other hand, we likely want this method to give _more_ positives than for regular `BlankNode`s.

I expect a `BlankNodeSkolemizable` A to equal B iff both are `BlankNodeSkolemizable` instances and they share the same `skolemized` (regardless of `value`).

So we could define `equals` such that it fails for all non-`BlankNodeSkolemizable` instances (including any other `BlankNode` implementation). Then we'd need to ensure that the reverse is also true. We can accomplish that by ensuring that `value` is set to a unique value every time. (For instance, UUID, or just ensuring that whatever thing generates blank node `value`s never reuses. N3.js does that by default, for instance.) Which you seem to be doing, so fine.

Reply:
I initially had an `instanceof BlankNodeSkolemizable` check here, which I removed to ensure symmary as you've mentioned.

Since `equals` is defined on the term, such an `instanceof` check can only be used in 50% of the cases. And since we already make sure `value` is never reused, I don't see any benefit in adding such a check, just confusion and weird edge-cases.

Reply:
But then I'm strongly wondering about the necessity of `skolemized`. Also as per my "human-accessible" comment below: https://github.com/comunica/comunica/pull/624#discussion_r393927250